encoder/vaapiencoder_vp9.cpp

```diff
@@ -0,0 +1,323 @@
+    }
+    else {
+        picParam->refresh_frame_flags = 0x01; //refresh last frame
+        picParam->pic_flags.bits.frame_type = kInterFrame;
+
+        ReferenceQueue::const_iterator it = m_reference.begin();
+
+        for (uint32_t i = 0; it != m_reference.end(); ++it, i++)
+            picParam->reference_frames[i] = (*it)->getID();
+
+        it = m_reference.begin();
+
+        // last/golden/alt is used as reference frame. L0 forward
+        picParam->ref_flags.bits.ref_frame_ctrl_l0 = 0x7;
+
+        // golden and alt are last KeyFrame
+        // last is last decoded frame
+        picParam->ref_flags.bits.ref_last_idx = 0;
+        picParam->ref_flags.bits.ref_gf_idx = 1;
+        picParam->ref_flags.bits.ref_arf_idx = 2;
+        picParam->pic_flags.bits.frame_context_idx = 0;
+    }
+
+    picParam->frame_width_src = width();
+    picParam->frame_height_src = height();
+    picParam->frame_width_dst = width();
+    picParam->frame_height_dst = height();
+
+    picParam->pic_flags.bits.show_frame = 1;
+
+    picParam->luma_ac_qindex = kDefaultQPValue;
+    picParam->luma_dc_qindex_delta = 1;
+    picParam->chroma_ac_qindex_delta = 1;
+    picParam->chroma_dc_qindex_delta = 1;
+
+    picParam->filter_level = kDefaultFilterLevel;
+    picParam->sharpness_level = kDefaultSharpnessLevel;
+
+    return true;
+}
+
+bool VaapiEncoderVP9::fill(
+    VAEncMiscParameterTypeVP9PerSegmantParam* segParam) const
+{
+    return true;
+}
+
+bool VaapiEncoderVP9::ensureSequence(const PicturePtr& picture)
+{
+    if (picture->m_type != VAAPI_PICTURE_I)
+        return true;
+
+    VAEncSequenceParameterBufferVP9* seqParam;
+    if (!picture->editSequence(seqParam) || !fill(seqParam)) {
+        ERROR("failed to create sequence parameter buffer (SPS)");
+        return false;
+    }
+    return true;
+}
+
+bool VaapiEncoderVP9::ensurePicture(const PicturePtr& picture,
+                                    const SurfacePtr& surface)
+{
+    VAEncPictureParameterBufferVP9* picParam;
+
+    if (!picture->editPicture(picParam) || !fill(picParam, picture, surface)) {
+        ERROR("failed to create picture parameter buffer (PPS)");
+        return false;
+    }
+    return true;
+}
+
+bool VaapiEncoderVP9::ensureQMatrix(const PicturePtr& picture)
+{
+    VAEncMiscParameterTypeVP9PerSegmantParam* segmentParam;
+
+    if (picture->m_type != VAAPI_PICTURE_I)
+        return true;
+
+    if (!picture->editQMatrix(segmentParam) || !fill(segmentParam)) {
+        ERROR("failed to create qMatrix");
+        return false;
+    }
+    return true;
+}
+
+bool VaapiEncoderVP9::referenceListUpdate(const PicturePtr& pic,
+                                          const SurfacePtr& recon)
+{
+
+    if (pic->m_type == VAAPI_PICTURE_I) {
+        m_reference.clear();
+        m_reference.insert(m_reference.end(), kMaxReferenceFrames, recon);
+    }
+    else {
+        m_reference.pop_front();
+        m_reference.push_front(recon);
```

Comment:
Hi @chivakker 
The code is not match you descriptions.
For first P frame, the reference list is
KKKKKKKK
You use first 3 reference, it will be KKK
For second P frame, the reference list is 
KKKKKKKP,
You still use first 3 reference, it will be KKK too.